[Merged by Bors] - feat(linear_algebra/affine_space): affine spaces

jsm28 opened the pull request:
Define (very minimally) affine spaces (as an abbreviation for
add_torsor in the case where the group is a vector space), affine
subspaces, the affine span of a set of points and affine maps.

jsm28 committed:
feat(linear_algebra/affine_space): affine spaces

Define (very minimally) affine spaces (as an abbreviation for
add_torsor in the case where the group is a vector space), affine
subspaces, the affine span of a set of points and affine maps.

robertylewis commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,206 @@
+/-- `affine_space` is an abbreviation for `add_torsor` in the
+case where the group is a vector space. -/
+@[nolint unused_arguments]
+abbreviation affine_space (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V]
```
The only instance that's really needed here is `[add_group V]`, which means every time this declaration is used, Lean will run two unnecessary type class searches. Is there a real reason to require them in this definition? Presumably they're needed in declarations below, and of course it's fine to require them there.

jsm28 replied:
Everything that only requires an `add_group` should already be in `algebra.add_torsor`. The definitions in this file depend on `submodule.span` or `linear_map`; some of them might work for modules in place of vector spaces, but not more generally than that.

jsm28 replied:
(Thus, this definition simply corresponds to one of the conventional definitions of what an affine space is.)

jsm28 replied:
I think the essential question here is: is it appropriate to have the `abbreviation affine_space` at all (because if so, it should have the arguments appropriate to the concept of an affine space), or should this file just comment on what an affine space is without defining `affine_space`, and then define other things such as `affine_subspace` and `affine_map` in terms of `add_torsor` together with whatever the requirements on `V` are for each individual definition?

jcommelin replied:
I think it's ok to have the definition. You could nevertheless "cheat" and drop the `field` assumption when it's not needed, just to make life easier for lean. Most theorems and lemmas will still have the assumption.

jsm28 replied:
In fact all the definitions and proofs here work for a `module` over a `ring` in place of a vector space (although there may well be some in future that don't), and it seems that affine spaces over a module are a real mathematical concept (although not mentioned on Wikipedia). One proof fails with a `semimodule` over a `semiring` (using current mathlib master); I think the result `vsub_mem_vector_span_of_mem_span_points_of_mem_span_points` is still true in that case, but would need a different, more complicated proof as `neg_one_smul` is no longer valid in the `semimodule` case (and I'm not sure affine spaces over semimodules are actually useful objects).

kim-em replied:
I agree that it's best to keep the definition, but without `[field k]`. You can make the doc-string more explicit about the fact that for all uses of the definition, we will need `[field k]`. It would be fine to be even more explicit, saying "but we omit the typeclass `[field k]` in the type synonym itself to simplify typeclass search".

jsm28 replied:
I've removed `[field k]` and thus `[vector_space k V]` from the `affine_space` definition, expanding the comment as suggested. It still needs `@[nolint unused_arguments]` unless we also remove `(k : Type*)` as an argument, should that be removed as well?

kim-em replied:
No, `k` itself should stay. It's pretty common for type synonyms to need a `@[nolint unused_arguments]`.

robertylewis commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,206 @@
+  use hp,
+  use 0,
+  exact and.intro (submodule.zero _) (zero_vadd V p).symm
+end
```
```suggestion
lemma span_points_nonempty_of_nonempty {s : set P} :
  s.nonempty → (span_points k V s).nonempty 
| ⟨p, hp⟩ := ⟨p, p, hp, 0, submodule.zero _, (zero_vadd V p).symm⟩
```

Proofs below can be similarly condensed.

robertylewis replied:
Even without this change, note that the middle lines could just be `use [p, p, hp, 0]`.

robertylewis commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,206 @@
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction.carrier)
+
+/-- The whole affine space as a subspace of itself. -/
+def affine_subspace.of_affine_space : affine_subspace k V P :=
```
Is this name following a pattern from somewhere else? I'm not entirely up on the naming conventions in this corner of the library, but this sounds a little funny. We're already in the `affine_space` namespace, and projection notation won't be useful. `univ_affine_subspace` or something like that?

jsm28 replied:
Renamed as suggested.

jsm28 committed:
Update src/linear_algebra/affine_space.lean

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

jsm28 committed:
Rename as per review.

jsm28 committed:
Shorten some proofs.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+open add_torsor
+
+variables (k : Type*) (V : Type*) {P : Type*} [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
```
```suggestion
variables [S : affine_space k V P]
```
Not very important, but can be useful for `grep` etc...

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+/-- The points in the affine span of a (possibly empty) set of
+points. -/
+def span_points (s : set P) : set P :=
+{p | ∃ p1 ∈ s, ∃ v ∈ (vector_span k V s).carrier, p = v +ᵥ p1}
```
You shouldn't need to write `.carrier` (and it will be annoying down the road if you do, because there are no `simp`-lemmas for it). You may need to write `\exists v : P, h : v \in (vector_span k V s)`.

jsm28 replied:
Removed all uses of `.carrier` for vector subspaces.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction.carrier)
+
+/-- The whole affine space as a subspace of itself. -/
+def univ_affine_subspace : affine_subspace k V P :=
```
```suggestion
def affine_subspace.univ : affine_subspace k V P :=
```
Just like `set.univ` and `finset.univ`.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+
+/-- The points in the affine span of a (possibly empty) set of
+points. -/
+def span_points (s : set P) : set P :=
```
Why not call this `affine_span`?

jsm28 replied:
`affine_span` is what I think of as the main function here, that gives you an `affine_subspace` but thus requires the set to be nonempty. Expanded the comment to refer to `affine_span`.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+subspace containing those points. (Actually defined here in terms of
+spans in vector spaces.) -/
+def affine_span (s : set P) (h : s.nonempty) : affine_subspace k V P :=
+{ carrier := span_points k V s,
```
If `span_points` is "deprecated" from now on, I think you should flag that in a comment at the definition of `span_points`.

jsm28 replied:
I don't think it's deprecated, but in most cases (if you know the set is nonempty), you probably want an `affine_subspace` and so probably want this function instead. I expanded the command on `span_points`.

jcommelin replied:
Ok, if `span_points` is not deprecated, then you probably want a lemma that says that the `coe` of `affine_span` to a set is `span_points`.
There should probably also be a `has_mem` instance for affine subspaces. Besides that, a lemma that says that elements of `s` are contained in `affine_span s`.

jsm28 replied:
Added that coercion, `has_mem` instance and lemmas.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+/-- An `affine_subspace k V P` is a subset of an `affine_space k V P`
+that has an affine space structure induced by a corresponding subspace
+of the `vector_space k V`. -/
+structure affine_subspace :=
```
Are you planning to define a lattice structure on `affine_subspace`? Probably `affine_span` is part of a galois connection, which might give you a bunch of "trivial lemmas" in a trivial way. Maybe a future PR?

jsm28 replied:
There's the complication that the intersection of two affine subspaces may be empty (but allowing affine subspaces to be empty would result in other complications, e.g. the choice of direction for such a subspace and defining its dimension). I suppose you could define e.g. a `semilattice_sup` on `affine_subspace`, and a lattice structure on `with_bot (affine_subspace k V P)`.

jcommelin replied:
Hmm, you are right. Let's ignore the lattice for the time being.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+
+/-- An `affine_map k V1 P1 V2 P2` is a map from `P1` to `P2` that
+induces a corresponding linear map from `V1` to `V2`. -/
+structure affine_map :=
```
I think it would be better to `extends linear_map k V1 V2`.

jsm28 replied:
That would mean `affine_map.to_fun` is the `to_fun` of the `linear_map`, and the function from points to points is called something else. Is that's what intended?

kim-em replied:
That doesn't sound so good.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,180 @@
+    refl
+  end
+}
+
```
It would be good to add a `coe_to_fun` instance, and some "standard" simp lemmas. Take a look just after the definition of `ring_hom` of `linear_map` to get an idea.

bryangingechen replied:
(Also the `}` should be on the same line as `end`.)

jsm28 replied:
Added that coercion and some lemmas.

jcommelin reviewed (commented):
Looks pretty good. I left some comments.

jsm28 committed:
Update src/linear_algebra/affine_space.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

jsm28 committed:
Update src/linear_algebra/affine_space.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

jsm28 committed:
Changes from review.

jsm28 committed:
Add coercion from affine maps to functions.

jcommelin commented on src/linear_algebra/affine_space.lean:
```diff
@@ -158,6 +158,33 @@ structure affine_map :=
+  conv_lhs {
+    rw [←vsub_vadd V1 p1 p2, affine_map.map_vadd, vadd_vsub]
+  }
+end
```
```suggestion
by conv_lhs { rw [←vsub_vadd V1 p1 p2, affine_map.map_vadd, vadd_vsub] }
```

jsm28 committed:
Add coercion from subspace to set, has_mem and related lemmas.

jsm28 committed:
Update src/linear_algebra/affine_space.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

kim-em commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,234 @@
+# Affine spaces
+
+This file defines affine spaces and subspaces and affine maps and the
+affine span of a set of points.
```
```suggestion
This file defines affine spaces and subspaces, affine maps, and the
affine span of a set of points.
```

jsm28 committed:
Update src/linear_algebra/affine_space.lean

Co-authored-by: Scott Morrison <scott@tqft.net>

jsm28 committed:
More changes from review.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into affine_space

kim-em commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,235 @@
+by conv_lhs { rw [←vsub_vadd V1 p1 p2, affine_map.map_vadd, vadd_vsub] }
+
+/-- Identity map as an affine map. -/
+def affine_map.id : affine_map k V1 P1 V1 P1 :=
```
You need a `@[simp]` lemma saying what `affine_map.id` does when applied to a point.

jsm28 replied:
Added `affine_map.id_apply`.

kim-em commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,235 @@
+instance : inhabited (affine_map k V1 P1 V1 P1) := ⟨affine_map.id k V1 P1⟩
+
+/-- Composition of affine maps. -/
+def affine_map.comp (f : affine_map k V2 P2 V3 P3) (g : affine_map k V1 P1 V2 P2)
```
idem

jsm28 replied:
Added `affine_map.comp_apply`.

kim-em commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,235 @@
+
+/-- An `affine_map k V1 P1 V2 P2` is a map from `P1` to `P2` that
+induces a corresponding linear map from `V1` to `V2`. -/
+structure affine_map :=
```
It would be nice to have an `@[ext]` lemma here, stating that to check two affine maps are equal, it suffices to check that the coercions to functions agree. (And in particular that one doesn't need to compare the `linear` fields.)

jsm28 replied:
Added `affine_map.ext`.

kim-em commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,235 @@
+result of subtracting the result of the affine map on those two
+points. -/
+@[simp] lemma affine_map.map_vsub (f : affine_map k V1 P1 V2 P2) (p1 p2 : P1) :
+  f p1 -ᵥ f p2 = f.linear (p1 -ᵥ p2) :=
```
It's not so obvious to me that this should be a `@[simp]` lemma, but I could be persuaded either way.

jsm28 replied:
Removed `@[simp]` from this lemma.

jcommelin replied:
If anything, I think the other direction could be a simp lemma (just like we simp `f (x + y) = f x + f y` for additive maps.

jsm28 committed:
More changes from review.

jsm28 committed:
More use of namespaces, to avoid instance name conflicts.

Aims to fix "invalid object declaration, environment already has an
object named 'set.has_coe'" from linting generated all.lean.

urkud commented on src/linear_algebra/affine_space.lean:
```diff
@@ -0,0 +1,267 @@
+(direction : submodule k V)
+(nonempty : carrier.nonempty)
+(add : ∀ (p : P) (v : V), p ∈ carrier → v ∈ direction → v +ᵥ p ∈ carrier)
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction)
```
Here is an alternative approach. I'm not sure which one is better but I think that we should discuss the pros and cons of each approach.

1. Define `affine_subspace` using the following axioms:
```lean
structure affine_subspace (k : Type*) (V : Type*) (P : Type*) [ring k] [add_comm_group V]
    [module k V] [affine_space k V P] :=
(carrier : set P)
(add_mem' : ∀ p₁ p₂ p₃ ∈ carrier, (p₂ -ᵥ p₁) +ᵥ p₃ ∈ carrier)
(smul_mem' : ∀ (c : k) (p₁ p₂ ∈ carrier), c • (p₂ -ᵥ p₁) +ᵥ p₁ ∈ carrier)
```

These axioms are equivalent to yours with two differences:

1. I don't include `direction` in the data (though it can be reconstructed).
2. I don't require the carrier to be nonempty.

The main "pro" of this approach is that the affine subspaces defined this way form a complete lattice. In particular, the intersection of two affine subspaces is always an affine subspace, you can define `affine_span s = Inf {t : affine_subspace k V P | s ⊆ (t : set P)}` and have a `galois_connection` between `set P` and `affine_subspace k V P`.

As for `direction`, there are a few different possibilities:

1. define `direction' : with_bot (submodule k V)` to be `bot` for the empty affine space or the set of `p1 -ᵥ p2` otherwise.
2. define `direction s [nonempty (s : set P)] : submodule k V`;
3. define `direction s : submodule k V` to be the zero submodule for the empty affine space, otherwise as in 1.

Of course, you can add two of these definitions and prove that they agree (look at `polynomial.degree` vs `polynomial.nat_degree`).

kim-em replied:
This does sounds pretty nice!

@jsm28, you're definitely allowed to include "sounds good, but I don't feel up to attempting this now" in your answer. :-)

jsm28 replied:
I think that may be my shorter answer, then.

As a longer answer, I suspect there might be use for both types eventually, but the one not requiring `nonempty` (i.e. sets of points containing all affine combinations) might best be called something other than `affine_subspace`. That one gives the abstract benefit of the lattice structure, but results in various other complications, and I'm less sure of having practical uses for it (but it may be hard to tell what complications arise with each approach without a lot more implemented that builds on affine subspaces).

Anything involving dimensions is likely to be more complicated with empty subspaces, because an empty space should have dimension -1, but dimensions are also naturally expected to be a nat, or a cardinal in the infinite-dimensional case, and to be the same as the dimension of the `submodule` that is the `direction`; dimension 0 is naturally equivalent to having a single point, so if you special-case empty subspaces as also having dimension 0 to keep the types convenient you lose that equivalence. Two subspaces should be parallel iff their directions are equal, but with empty subspaces that doesn't work if the direction has type `submodule` which would make empty spaces parallel to those with a single point. Without empty subspaces, two subspaces are equal iff they have nonempty intersection and the same direction. With empty subspaces, it's no longer the case that any subspace has an affine frame, and a subspace of dimension n only necessarily has a barycentric frame of n+1 points if your definition allows dimension -1 (and various definitions around barycentric coordinates would probably need special-casing for empty subspaces). Empty subspaces can't be converted to affine spaces on a subtype of points, whereas nonempty ones can.

As a general implementation principle, I hope that most definitions and proofs around affine subspaces and spans can do their main work through the corresponding definitions and results for submodules and spans in modules, rather than needing to repeat any logic that's already present in the module case.

I do know that for geometry I'll want types that build on top of nonempty subspaces by requiring them to have a given dimension, so that e.g. types `line k V P` or `plane k V P` can be defined to `affine_subspace_n_dim k V P 1` or `affine_subspace_n_dim k V P 2` or similar (and whenever a 0-dimensional subspace arises, it will probably be more convenient to convert it back to a point of type `P`).

I don't right now have any working code outside of this PR that builds on affine subspaces. (I do have code building on affine maps, at a fairly trivial level - constructing an `affine_map` from an isometry of Euclidean spaces.)

urkud replied:
Just to be clear:

* This was a suggestion, i.e. I won't insist at least until I'll need this for some good reason (most probably I won't have a reason to use this part of `mathlib` in the next several months); if I'll have a good reason, I'll come again and will be ready to do the refactor.
* I do not suggest proving facts about the "bad" empty subspace when it's not convenient. Just add `[nonempty (s : set P)]` or `(hs : (s : set P).nonempty)` to any `def`/`lemma`.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into affine_space

jcommelin reviewed (commented):
lgtm, shall we get this merged? @urkud @semorrison

jcommelin reviewed (approved):
Thanks :tada:

bors merge
